module/data/activity/attack-data.mjs

```diff
@@ -223,4 +223,84 @@ export default class AttackActivityData extends BaseActivityData {
 
     return { data, parts };
   }
+
+  /* -------------------------------------------- */
+
+  /**
+   * @typedef {AttackDamageRollProcessConfiguration} [config={}]
+   * @property {"oneHanded"|"twoHanded"|"offHand"|"thrown"} mode  Attack mode.
+   */
+
+  /**
+   * Get the roll parts used to create the damage rolls.
+   * @param {Partial<AttackDamageRollProcessConfiguration>} [config={}]
+   * @returns {AttackDamageRollProcessConfiguration}
+   */
+  getDamageConfig(config={}) {
+    const rollConfig = super.getDamageConfig(config);
+
+    // TODO: Handle ammunition
+    // If base part is present, add ammunition physical properties to it
+    // If ammo has its own damage, create a new damage part of it
+
+    // const ammo = this.hasAmmo ? this.actor.items.get(this.system.consume.target) : null;
+    // if ( ammo ) {
+    //   const properties = Array.from(ammo.system.properties).filter(p => CONFIG.DND5E.itemProperties[p]?.isPhysical);
+    //   if ( this.system.properties.has("mgc") && !properties.includes("mgc") ) properties.push("mgc");
+    //   const ammoConfigs = ammo.system.damage.parts.map((([formula, type]) => ({ parts: [formula], type, properties })));
+    //   if ( ammo.system.magicalBonus && ammo.system.magicAvailable ) {
+    //     rollConfigs[0].parts.push("@ammo");
+    //     properties.forEach(p => {
+    //       if ( !rollConfigs[0].properties.includes(p) ) rollConfigs[0].properties.push(p);
+    //     });
+    //     rollData.ammo = ammo.system.magicalBonus;
+    //   }
+    //   rollConfigs.push(...ammoConfigs);
+    // }
+
+    if ( this.damage.critical.bonus ) {
+      rollConfig.critical ??= {};
+      rollConfig.critical.bonusDamage ??= this.damage.critical.bonus;
+    }
+
+    return rollConfig;
+  }
+
+  /* -------------------------------------------- */
+
+  /** @inheritDoc */
+  _processDamagePart(damage, rollConfig, rollData) {
+    if ( !damage.base ) super._processDamagePart(damage, rollConfig, rollData);
+
+    // Swap base damage for versatile if two-handed attack is made on versatile weapon
+    if ( this.item.system.isVersatile && (rollConfig.mode === "twoHanded") ) {
+      const versatile = this.item.system.damage.versatile.clone();
+      versatile.base = true;
+      versatile.denomination ||= damage.steppedDenomination();
+      versatile.number ||= damage.number;
+      versatile.types = damage.types;
+      damage = versatile;
+    }
+
+    const roll = super._processDamagePart(damage, rollConfig, rollData);
+    roll.base = true;
+
+    if ( this.item.type === "weapon" ) {
+      // Ensure `@mod` is present in damage unless it is a off hand attack
+      if ( (rollConfig.mode !== "offHand") && !roll.parts.some(p => p.includes("@mod")) ) roll.parts.push("@mod");
```

Comment:
The modifier is still added for off-hand attacks if it's negative.

That aside, it's probably more ergonomic to have an `off-hand` weapon property that does this automatically, rather than having to choose the 'off-hand' mode every time you roll with the weapon. This doesn't work for versatile or thrown since in those cases you have a single weapon with different ways to attack, but with dual-wielding you have two separate weapons that can have separate properties no problem.

Reply:
Just to clarify: not suggesting we add an off-hand property in this PR. Either we decide the `mode` solution is good, or if we prefer the `properties` solution then we remove this bit of code and defer to later.

Reply:
I was thinking of storing the last used mode that you select when you attack. That will make it easy to change if you decide to use your normally off-hand weapon as your main weapon without having to open up the item sheet to change the settings, while still making it fast to activate.

Reply:
Yeah that makes sense, we need some solution for this for the versatile case anyway, since selecting that option every time is not tenable.